### ISSUE TITLE
Bug: [no-magic-numbers] False positive when using `as const`

### ISSUE THREAD

**Issue by AviDuda** [typescript-eslint/typescript-eslint#6558]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.5&sourceType=module&code=MYewdgzgLgBAngUwIYCcYF4YEYBMBuAKANElkVQEEIBhcaDbHGJCGE6Q9s5FVzAbSwAaGDhEBmALrNWXPEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFsBDAc0plaTWNwBGiaKgyQA7r2hNI4MAF8QaoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const year = 12 as const;
const years = [1, 2, 3] as const;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-magic-numbers": "warn",
  },
};
```

### Expected Result

No warning/error when using `as const`, as it doesn't warn when omitting the const assertion.
There are pretty much always some magic numbers (like months in the example above), and it's better when they're defined `as const`, which makes it easier to see the real value of the magic constant instead of a generic `number`.

### Actual Result

Warnings like `No magic number: 12.`.

### Additional Info

I wouldn't mind this as an extra option, for example `ignoreAsConst`.

**Comment by bradzacher** [comment]
For the array case - I don't think this case should be covered as its not covered by the base lint rule.

As for the non-array case - the `as const` is entirely superfluous - [TS by default will type a `const` variable as narrowly as it can when it comes to literal values.](https://www.typescriptlang.org/play?#code/LAKAxg9gdgzgLgAgJ4FMCGAnBBeBBGAJgG5QB6UhSygPQH5RRJZFVMBBGAYWnh3wIRoYCJvBIhyVGvRBA)
So I'm not sure that it's worth adding an exception here.

Unless there's some cases I don't know about?

**Comment by JoshuaKGoldberg** [comment]
It's been ~1.5 months without a response, closing this issue. Please post back or file a new issue (following the template & including a correct example) if you want action to be taken. Thanks!
